### PR TITLE
Add checkpoint message (non-)inclusion verification

### PR DESCRIPTION
**Pull request by Hyodar**
This PR adds utility methods for checkpoint message inclusion verification to the `SFFLTaskManager` contract. It does not yet include the checkpoint tree generation on operators, this is work for another PR.

Additional context:
* An SMT was used, as opposed to an MPT - this was the initially thought approach. SMTs are a lot simpler and cheaper to verify, while also offering KV-like capabilities such as an MPT. The only con found was that there is less existing support for those. Still, a [suitable go library](https://github.com/pokt-network/smt) was found, and in terms of Solidity it was preferrable to implement the verification algorithm manually as libraries are not well-maintained and the root computing algorithm is really simple as well - quite similar to a normal Merkle Tree's.
* This is clear when checking it, but in case it's not well understandable quickly - SMT non-inclusion proofs are actually null-inclusion proofs, as an SMT abstractly includes every possible key as null by default. So, in terms of our uses, in order to verify inclusion, you need to specify the message and pass a valid inclusion proof for the non-null message digest - and in order to verify non-inclusion, you need to specify the message and pass a valid inclusion proof for a null leaf or a leaf that differs from the message digest.

The PR also resulted in a big renaming in order to properly move the `OperatorSetUpdate` library. It can be filtered out for a cleaner diff by filtering commits from [a69e175](https://github.com/NethermindEth/near-sffl/pull/53/commits/a69e175d72130182ffd5146dbfc8a98f038689f2) to [84b776b](https://github.com/NethermindEth/near-sffl/pull/53/commits/84b776bd9779e99a1a53888e2083efcb3b304415).

### CODE REVIEW

File: contracts/evm/src/base/utils/RollupOperators.sol [hunk 1]
```diff
@@ -11,7 +11,7 @@ import {BN254} from "eigenlayer-middleware/src/libraries/BN254.sol";
  * @dev The operator set is an alternative representation of the AVS' original
  * operator set, as it assumes a one-quorum one-weight based voting
  */
-library Operators {
+library RollupOperators {
```
Comment (taco-paco) [inline contracts/evm/src/base/utils/RollupOperators.sol:14]:
```suggestion
library AvsOperators {
```

Those operators both on Rollups and L1 so `RollupOperators` feels a bit misleading.

**Review by taco-paco — APPROVED** [review]
looks good

**Comment by Hyodar** [comment]
Putting this PR on hold for now - some of the Go SMT implementations do things a bit differently between themselves, which leads to some changes in the verification algorithm.

**Comment by Hyodar** [comment]
Updated the verifier and related tests for compatibility with https://github.com/pokt-network/smt.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        if (depth != 256) {
+            require(proof.index < 2 ** depth, "Invalid index");
+        }
+
+        bytes32 leaf = proof.leaf;
+
+        uint256 siblingIdx;
+        for (uint256 i = 0; i < depth; i++) {
+            bytes32 sibling = ((proof.bitmask >> i) & 1) == 1 ? proof.sideNodes[siblingIdx++] : bytes32(0);
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Done in the new verifier implementation

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+     * through Yul, properly using the scratch area
+     * @param left First hash function argument
+     * @param right Second hash function argument
+     * @return result Hash function result
+     */
+    function _hash(bytes32 left, bytes32 right) private pure returns (bytes32 result) {
+        if (left != bytes32(0) || right != bytes32(0)) {
+            /// @solidity memory-safe-assembly
+            assembly {
+                mstore(0x00, left)
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Not relevant anymore, but yes it is in this case

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     * @notice Computes the SMT root through an SMT (non-)inclusion proof.
+     * @param depth SMT depth
+     * @param proof SMT proof
+     * @return SMT root
+     */
+    function computeRoot(uint256 depth, Proof calldata proof) internal pure returns (bytes32) {
+        require(depth <= 256, "Depth should be at most 256");
+
+        if (depth != 256) {
+            require(proof.index < 2 ** depth, "Invalid index");
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Yes, but bitmask is already a `uint256` and only the relevant bits are considered. If there are more set bits than side nodes, it will lead to an index out of bounds exception when reading the side node from `sideNodes`, so should not be an issue.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: As mentioned in the comment above, if there are more set bits than expected, we'll get an exception either way. I don't think this extra check is worth the extra gas.

[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Also, we could remove `bitmask`, but I think it's a good way to make the proof more compact (this is the difference between a compact and a non-compact proof) - instead of making the caller send a 256-long side nodes array at max, the array length with `bitmask` is <= 256 side nodes, which should potentially save a considerable amount of gas.

File: contracts/evm/src/base/utils/RollupOperators.sol (hunk 1 above)
Comment (Hyodar) [reply contracts/evm/src/base/utils/RollupOperators.sol:14]: I actually think this makes things a bit more confusing. This is about the operator definition for rollups, since there is another definition for mainnet. `AvsOperators` imo doesn't differentiate between the two. Wdyt?

Comment (Hyodar) [reply contracts/evm/src/base/utils/RollupOperators.sol:14]: The library is used in both because the operator set update message, which is meant to be passed to rollups, is based on the rollup operators, but it's verifiable in mainnet checkpoints.
We need to differentiate them because even though the meaning is similar the formats and assumptions are different. The operator format for mainnet is determined by the EL core contracts, whereas the rollup operator format is a compact version of it.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        } else {
+            currentHash = _hashNode(hashBuffer, LEAF_NODE_PREFIX, path, proof.value);
+        }
+
+        uint256 sideNodeIndex = 0;
+        for (uint256 i = 0; i < proof.numSideNodes; i++) {
+            uint256 bit = proof.numSideNodes - i - 1;
+            bytes32 sideNode = ((proof.bitMask & (1 << i)) != 0) ? DEFAULT_VALUE : proof.sideNodes[sideNodeIndex++];
+
+            if ((uint256(path) >> (255 - bit)) & 1 == 0) {
```
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:85]: This
```
if ((uint256(path) >> (255 - bit)) & 1 == 0)
```
is equivalent to this
```
if ((uint256(path) >> (256 - proof.numSideNodes + i)) & 1 == 0)
```
Here we basically first shift on 256 - proof.numSideNodes removing junk bits. And on next iterations shift only on 1 bit.

So I would propose rewriting function as follows.

File: contracts/evm/src/base/utils/RollupOperators.sol (hunk 1 above)
Comment (Hyodar) [reply contracts/evm/src/base/utils/RollupOperators.sol:14]: No, I'm not talking about context of use, I'm talking about actual data formats. There's not the same Operator data format in EL's core contracts. RollupOperators.Operator is indeed only used in the rollup contracts. The way verification is done and such is completely different between both.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+     * Ideally this could also use free memory (in terms of free memory
+     * pointer), but memory may not be properly cleaned afterwards and it could
+     * be (wrongly) considered as clean memory when actually being allocated
+     * @param hashBuffer Memory area to be used for serialization
+     * @param prefix Node prefix for hashing
+     * @param left Left value
+     * @param right Right value
+     * @return result Result of `keccak256(prefix || left || right)`
+     */
+    function _hashNode(bytes32[3] memory hashBuffer, uint8 prefix, bytes32 left, bytes32 right)
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:111]: This is already the case - this just passes a pointer to a previously reserved memory slot. We can't really use the scratch area in this case though, it's only 64 bytes.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:85]: Makes sense. I was trying to keep it similar to Pokt's implementation as much as possible, but this seems like a good change

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    function verifyProof(bytes32 root, Proof calldata proof) internal pure returns (bool) {
+        require(proof.sideNodes.length <= MAX_DEPTH && proof.numSideNodes <= MAX_DEPTH, "Side nodes exceed depth");
+
+        bytes32 path = keccak256(abi.encodePacked(proof.key));
+        bytes32[3] memory hashBuffer;
+
+        bytes32 currentHash;
+
+        if (proof.value == DEFAULT_VALUE) {
```
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:67]: I thought I'd still agree with you on that, but started trying it out and honestly. besides gas considerations (minimal), I feel like it's not enough nesting to justify the encapsulation, and it's like 10 lines atm. One more place to read in order to understand what's happening, and this is only used in one situation to determine the leaf we're trying to prove. I'd still be up to keeping it as it is.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Done in [7b9dcbe](https://github.com/NethermindEth/near-sffl/pull/53/commits/7b9dcbe4b581e7f9d7d6e0e103add51aacec0c3e)

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:67]: This whole logic just resolves current hash. It's a separate logic unit which is part of algorithm you implemented which just complicates overall already not easy to read algorithm. 

Imo resulting code [here](https://github.com/NethermindEth/near-sffl/pull/53#discussion_r1548071665 ) looks much more readable. And if one decides to check how actually currentHash computed he can just navigate to the function.

[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:67]: As one who actually had to read without a lot of context on algorithm itself, initial version of this math trickery was a bit of challange to understand. Too much things going on: magical bitshifts, magical offsets, if elses on top.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (taco-paco) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:112]: I could see that its previously allocated memory) Was wondering if there anything similiar to static variables in solidity not to argument around? 

Are variables declared inside contract(not functions) only "storage" variables rr it can be "memory"?

**Review by taco-paco — APPROVED** [review]
Looks good. But I would still prefer to make verifyProof a bit more dummy friendly. For like of me

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:112]: Got what you mean. We can't do it in Solidity afaik - contract variables are storage-based, and there's no way to make a static variable so we didn't need to pass the pointer around.

File: contracts/evm/src/eth/utils/SparseMerkleTree.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol:67]: Let me try again and see how it looks. Usually MT verifiers have similar implementations, so that's maybe why it doesn't seem strange to me. In the end, what happens in this part is simply computing the leaf it's trying to prove based on `value` and `nonMembershipLeaf*`, which is the 'initial hash' before going to the inner nodes.

[the remark this answers is not in the document]
Comment (Hyodar) [reply contracts/evm/src/eth/utils/SparseMerkleTree.sol]: Done in [8e108e2](https://github.com/NethermindEth/near-sffl/pull/53/commits/8e108e24a2d9b1aabba48f652c0986771164130f). Also added `_computeProofRoot` on [8b822e6](https://github.com/NethermindEth/near-sffl/pull/53/commits/8b822e6a28e35d0ac64b477d570e65a76d8f263f).
